Light Filter Visualisation for 34.0 Maintenance Branch

mattigruener opened the pull request:
Hey guys,

first attempt at creating one of theses WIP PR's - @johnhaddon and @andrewkaufman, can you see this?

John, there are a few things in here that you won't like at all. We probably need to talk a bit about how you want me to clean this up. This PR includes all the light filter commits that have been merged into master already, btw.

I didn't find too much overlap between the Arnold filters and those developed by Arnold's competition so -for now- I have put everything into specific Arnold visualisers. The decay visualiser has some potential for additional abstraction though. We could impose a little interface there that enforces decay visualisers to implement a method that returns the knots that will be drawn. I thought visualising the ramp determining the decay via the knots might be the best way to handle all kinds of decays - at least people can see where those knots lie in space.

The biggest issue for me is how I handled the hand-off between OSL texture generation and GL (see OSLTextureReader). The texture loader bits of Cortex seem very much designed to only load images from disk, so what I've done here feels very hacky and goes wrong in all kinds of ways. All of these are obsolete if we tackle this differently, but I didn't want to change the API in cortex too much for this PR.

1. Replacing the parameter in the constructor - urg. I need to change the policy regarding file existence, though. I also had to remove some code in Cortex that does another explicit file existence check, but I won't add a PR for that for now.

2. I'm currently creating a new texture for every change, because the texture loader caches the ImagePrimitive if I don't change the file name. What I'm doing will eventually blow up, I guess, as the texture loader has no way of cleaning up after me.

For some bits of the code, it would be great if we could find a general location so that we can reuse it.

There are todo tags sprinkled throughout the code that are meant to draw attention to bits that I'm not happy with. I'm holding off on some of them so that I can discuss general direction with you first.

I like the OSL stuff in general, being able to put whatever OSL node I want into the network and see a somewhat quick feedback is nice.

Thanks :)

Matti.

PS: Andrew, I misspelled your github handle in the first iteration (Ctrl-Enter does not add a new bulletpoint, but finishes the editing... -.-),  - did you get a notification that you were mentioned?

johnhaddon commented on include/GafferArnoldUI/BarndoorVisualiser.h:
```diff
@@ -0,0 +1,65 @@
+	BarndoorVisualiser();
+	~BarndoorVisualiser();
+
+	virtual IECoreGL::ConstRenderablePtr visualise( const IECore::InternedString &attributeName, const IECore::ObjectVector *shaderVector, const IECore::Shader *lightShader, IECoreGL::ConstStatePtr &state ) const;
```
Remove virtual, add override?

mattigruener replied:
I guess that raises the question whether I should aim for getting this merged into 34 or just the next release that supports c++11? I had initially planned to eventually create a PR that converts some of the non-cpp11 stuff here into more modern cpp. I think I'm also using NULL instead of nullptr somewhere in the code...

johnhaddon replied:
Hmm, yeah, that's a tricky one. For what it's worth, my plan for the Encapsulate work is to use nice sensible C++11 idioms wherever I know I can backport fairly easily, but avoid anything that would be harder to backport. For instance, I'd absolutely use `override` because it's trivial to revert back, and I'd happily use `std::unique_ptr`, `std::shared_ptr` and `std::function` because it'd be straightforward to replace them with boost equivalents. But I'll be more wary of using lambdas or auto or range-for, because then the back porting is non-trivial.

johnhaddon commented on include/GafferSceneUI/StandardLightFilterVisualiser.h:
```diff
@@ -0,0 +1,67 @@
+
+/// Class for performing standard visualisations of light filters,
+/// by mapping shader parameters to features of the visualisation.
+class StandardLightFilterVisualiser : public LightFilterVisualiser
```
What's the reason for including this class? Unless I'm mistaken, it doesn't seem to be providing anything...

mattigruener replied:
You're not mistaken. I've just added it as a placeholder that could eventually address similarities between PrMan and Appleseed maybe? I'm happy to remove it. I really just mimicked the LightVisualiser side of things in anticipation that we'll want to add more renderers eventually.

johnhaddon replied:
Let's remove it for now, since you seemed pretty convinced that the metadata approach we took for StandardLightVisualiser isn't going to map well to the light filters.

johnhaddon commented on src/GafferArnoldUI/GoboVisualiser.cpp:
```diff
@@ -0,0 +1,339 @@
+		IECore::Shader *surfaceShader = new IECore::Shader( "Surface/Constant", "osl:shader" );
+		surfaceShader->parameters()["Cs"] = new StringData( slidemap->readable() );
+
+		IECore::ObjectVector *oslNetwork = new IECore::ObjectVector();
```
Memory leak? Always assign the result of `new` to a smart pointer of some sort - in this case ObjectVectorPtr. As of C++14 they're pretty much saying that you should never call `new`, because you can use `make_unique` and `make_shared` instead. I'm thinking about doing something to enforce the creation of RefCounted classes in a similar way, but haven't got there yet...

johnhaddon commented on include/GafferArnoldUI/BarndoorVisualiser.h:
```diff
@@ -0,0 +1,65 @@
+class BarndoorVisualiser : public GafferSceneUI::StandardLightFilterVisualiser
+{
+
+public :
```
Indentation.

johnhaddon commented on include/GafferArnoldUI/BarndoorVisualiser.h:
```diff
@@ -0,0 +1,65 @@
+namespace GafferArnoldUI
+{
+
+class BarndoorVisualiser : public GafferSceneUI::StandardLightFilterVisualiser
```
Where there's no need for anyone to directly work with a particular class, I'm tending to totally hide them these days, so that there'd just be a .cpp file and no header. This means we're free to change them without breaking binary compatibility, and there's simply less stuff for someone to look at when they're trying to find their bearings in the API. I wonder if these visualisers could be hidden that way?

mattigruener replied:
I guess we'd want to do that even though we need to bind it in python for the registering?

johnhaddon commented on src/GafferArnoldUI/BarndoorVisualiser.cpp:
```diff
@@ -0,0 +1,290 @@
+	return defaultValue;
+}
+
+void spotlightParameters( InternedString metadataTarget, const IECore::CompoundData *parameters, float &innerAngle, float &outerAngle, float &lensRadius )
```
I think I mentioned that we might provide some sort of LightAccessor class at some point, to provide an interface for setting and getting parameters like these. We could then use that in a LightSpreadsheet and light manipulators etc. Seems like we have another use case here. Rather than send this work off on a tangent though, perhaps we should just provide something like this function as a static method on the StandardLightVisualiser?

If we do that, I'd like to consider what the most appropriate arguments might be - it seems that `metadataTarget` and `parameters` are specific to the internal implementation of StandardLightVisualiser, and a future LightAccessor abstraction wouldn't use them either. Perhaps the most logical arguments are `lightAttributeName` and `const ObjectVector *lightShader`? Those are the highest-level representation of a light we have in our data flow, and require the least knowledge from the client (hence why they're also the arguments to `LightVisualiser::visualise()?`. Could that work?

johnhaddon commented on include/GafferArnoldUI/DecayVisualiser.h:
```diff
@@ -0,0 +1,65 @@
+
+	IE_CORE_DECLAREPTR( DecayVisualiser )
+
+		} // namespace GafferArnoldUI
```
Indentation.

johnhaddon commented on include/GafferArnoldUI/DecayVisualiser.h:
```diff
@@ -0,0 +1,65 @@
+class DecayVisualiser : public GafferSceneUI::StandardLightFilterVisualiser
+{
+
+public :
```
Indentation.

johnhaddon commented on include/GafferOSL/OSLTextureReader.h:
```diff
@@ -0,0 +1,78 @@
+namespace GafferOSL
+{
+
+class OSLTextureReader : public IECore::Reader
```
Ingenious! Unless I'm missing something, what you really want to do is pass an ImagePrimitive directly as a shader parameter to OpenGL? That isn't possible because shader parameters have to be Data-derived, but there is already a workaround that allows a CompoundData with the right format to be passed. The OpenGLShader [already uses that](https://github.com/GafferHQ/gaffer/blob/master/src/GafferScene/OpenGLShader.cpp#L174), so perhaps we could use the same here, and not need this OSLTextureReader class at all?

mattigruener commented on src/GafferArnoldUI/BarndoorVisualiser.cpp:
```diff
@@ -0,0 +1,290 @@
+
+	const IECore::CompoundData *filterShaderParameters = filterShader->parametersData();
+
+	float topLeft = parameterOrDefault( filterShaderParameters, "barndoor_top_left", 0.0f );
```
What I also should've mentioned is that it's a little unfortunate that we're currently hard-coding the defaults to be used in absence of parameters. It probably doesn't happen so much that we actually have to use the defaults, but if it does it would be much nicer to use the data that was read by ArnoldShader already? Do you think there's any value in getting the defaults from there or should be keep it simple like this?

johnhaddon replied:
I think there's an argument for getting it from the horse's mouth, but for now I'd be happy to go with the approach you have here, mostly just to avoid having to think about where it would belong otherwise. The visualisers shouldn't really have any dependencies on nodes like ArnoldShader, because the only thing they really need is the data that's generated, not the node that generated the data. We could use the `Ai*` API directly, but then we're into `AiUniverse` juggling territory. So we'd be looking at building a different stash of all the Arnold metadata somewhere else at startup - maybe even IECoreArnold? I don't think the potential issues justify the work right now...

mattigruener replied:
Yeah I was thinking about having a global cache somewhere that both the ArnoldNode and our visualisers could query, but we can still add that later :) Cool thanks. And thanks for teaching me a new expression :horse:

mattigruener commented:
Hey John,

I think I've addressed most of your feedback. I'm still using cpp < 11 because I don't think I have a good way to test my code if I rebase it onto current master. Couldn't get it to compile because of LRUCache issues and then moved back to my 34 branch.

The gobo visualiser will definitely need a cache so that we don't redo the osl evaluation all the time. I'm sure there is something in Cortex I could/should use for this one.

Thanks :)

johnhaddon commented on src/GafferArnoldUI/GoboVisualiser.cpp:
```diff
@@ -0,0 +1,282 @@
+		{
+			uWritable.push_back( (float)(x + 0.5f) / resolution );
+			vWritable.push_back( (float)(y + 0.5f) / resolution );
+			pWritable.push_back( V3f( x, y, 0.0f ) );
```
Just looking at the equivalent code in OSLImage, and we're adding 0.5 here too, so the points are at pixel centres. Probably not a big deal, but might be worth matching?

johnhaddon commented on src/GafferSceneUI/LightFilterVisualiser.cpp:
```diff
@@ -0,0 +1,228 @@
+	return l;
+}
+
+// const LightFilterVisualiser *standardLightFilterVisualiser()
```
Remove dead code.

johnhaddon commented on src/GafferSceneUI/LightFilterVisualiser.cpp:
```diff
@@ -0,0 +1,228 @@
+			continue;
+		}
+
+		// const IECore::Shader *lightShader = IECore::runTimeCast<const IECore::Shader>( lightShaderVector->members().back().get() );
```
Dead code.

johnhaddon commented on src/GafferSceneUI/LightFilterVisualiser.cpp:
```diff
@@ -0,0 +1,228 @@
+		// renderer:lightFilter only, though. It's therefore necessary to strip off
+		// the optional part
+		std::vector<std::string> strs;
+		boost::split( strs, it->first.string(), boost::is_any_of(":") );
```
Is this not the same as the `tokens` generated at line 140? It's not clear at first glance because this uses `it->first` and the other uses the `attributeName` alias, but I _think_ they're the same. There's a few other spots using `it->first` where `attributeName` would be clearer.

johnhaddon commented on src/GafferSceneUI/LightFilterVisualiser.cpp:
```diff
@@ -0,0 +1,228 @@
+		}
+		else
+		{
+			return NULL;
```
Should this be `continue`? It seems like we'll skip visualising anything else the first time we hit a missing visualiser otherwise?

johnhaddon commented on SConstruct:
```diff
@@ -715,7 +715,7 @@ libraries = {
 	"GafferArnoldUI" : {
 		"pythonEnvAppends" : {
-			"LIBS" : [ "GafferArnoldUI" ],
+			"LIBS" : [ "GafferArnoldUI", "GafferSceneUI" ],
```
No big deal, but if you get a chance, could you squash this into the barndoor commit with the other SConstruct modification? It wouldn't link on Mac otherwise, and it's nice to keep all commits buildable for git-bisect. Actually, on Mac I ended up needing this :

```
"GafferArnoldUI" : {
	"envAppends" : {
		"LIBS" : [ "IECoreGL$CORTEX_LIB_SUFFIX", "GafferOSL", "GafferSceneUI" ],
	},
	"pythonEnvAppends" : {
		"LIBS" : [ "GafferArnoldUI", "GafferSceneUI" ],
	},
},
```

On Mac, all symbols are checked when linking during the build, whereas on Linux it just wings it and only checks at runtime. That means that you can get lucky at runtime on Linux if something else has loaded a library that is needed but wasn't specified at build time, whereas the Mac wouldn't even let you build it...

johnhaddon commented on python/GafferArnoldUI/__init__.py:
```diff
@@ -48,5 +48,8 @@
 import CacheMenu
 
 from _GafferArnoldUI import *
+import GafferSceneUI
```
This causes a bit of namespace pollution whereby `GafferSceneUI` becomes available as `GafferArnoldUI.GafferSceneUI`. We have a [few tests](https://github.com/GafferHQ/gaffer/blob/master/python/GafferSceneTest/ModuleTest.py#L41) against that sort of thing, but they're not thorough enough to catch it.

I suggest we move the registrations to C++ anyway, which'll also allow us to do the total hiding of the classes I mentioned before...

mattigruener commented:
Thanks for the last round of feedback - I think I addressed all outstanding issues in the latest update. 

Now that we don't create bindings for the GafferArnoldUI lib anymore, I had to get that lib loaded through some other mechanism. I'm sure what I've done is not as intended, but I didn't know how you would want this handled... Initially, I created the GafferArnoldUI lib to mimic the GafferSceneUI side of things - I guess we could also put its contents into GafferArnold..?

johnhaddon commented on src/GafferArnoldUI/BarndoorVisualiser.cpp:
```diff
@@ -0,0 +1,261 @@
+const char *barndoorFragSource()
+{
+	return
+		"#if __VERSION__ <= 120\n"
```
Looks like we don't use the #define in this shader?

johnhaddon commented on src/GafferArnoldUI/BarndoorVisualiser.cpp:
```diff
@@ -0,0 +1,261 @@
+		""
+		"void main()"
+		"{"
+			"if( (int(gl_FragCoord.x) + int(gl_FragCoord.y)) % 2 == 0 )"
```
This isn't working on Mac. There is a `mod( float, float )` function though that does work. I've fiddled with that briefly to try to get a checkerboard but failed - maybe because gl_fragCoord is weird on a Retina display? Could you try to get something working with `mod()` at your end and then I'll test again here?

mattigruener replied:
I think I addressed all inline comments - will look at the comments about the visualisation itself next. 

I've used mod() now, does that work on your end?

johnhaddon commented on src/GafferArnoldUI/BarndoorVisualiser.cpp:
```diff
@@ -0,0 +1,261 @@
+	Imath::M44f trans;
+	switch( location )
+	{
+	case TOP:
```
ALL_CAPS is for macros and CamelCase is for enums...

johnhaddon commented on src/GafferArnoldUI/BarndoorVisualiser.cpp:
```diff
@@ -0,0 +1,261 @@
+
+} // namespace
+
+namespace GafferArnoldUI
```
I think everything can actually be in the anonymous namespace for all of these visualisers.

johnhaddon commented on src/GafferArnoldUI/GoboVisualiser.cpp:
```diff
@@ -0,0 +1,365 @@
+}
+
+typedef LRUCache<OSLTextureCacheKey, CompoundDataPtr> OSLTextureCache;
+OSLTextureCache g_oslTextureCache( getter, 10000 );
```
What's the memory usage associated with this limit? I make it this :

```
res * yres * channels * sizeof( float ) * cacheSizeLimit
512 * 512 * 3 * 4 * 10000
31457280000 bytes
30 gigabytes
```

Seems a bit much!

I think I'm right in saying that the visualisers are only called again a second time when the attributes have changed? In which case we would only need the cache to store the last value of the texture for each gobo, so it is effective when a non-texture parameter (like rotation) is changed. If the texture itself changes, there's little chance of it being in the cache anyway, and the user shouldn't mind a bit of a hit. I'd suggest a much lower limit - 100 say.

mattigruener replied:
I think storing gobo textures is a perfectly good use of 30GB of RAM! :P

No, what you suggest sounds good - one thing to keep in mind is that we allow not just texture reads but all kinds of osl manipulation. I've been playing with color grading the texture for example and caching the result for certain nodes being disabled seemed like a reasonable use of the cache to me too. The cache is definitely too large right now, though, and 100 seems like a good value. Sorry, I've been naively copying that number from the example you posted earlier without doing the maths.

johnhaddon commented on startup/gui/lightFilterVis.py:
```diff
@@ -0,0 +1,2 @@
+import ctypes
+ctypes.cdll.LoadLibrary( 'libGafferArnoldUI.so' )
```
This doesn't work on Mac, where the extension is ".dylib". I think all our modules contain some bindings at this point in time, but in the past when we've had the same issue, I've just made an empty module that is linked to the necessary library. `import` is the canonical way of getting functionality into Python, so I think it makes sense to continue to use it for this purpose.

mattigruener commented:
The last batch of changes should add the gobo transformations to the preview and make the triangle rendering a little more like what you want.

johnhaddon commented:
Sorry Matti, I don't think I'm going to get to looking at this properly today. Seems like we should be in good shape though, so do you fancy making a final PR on GafferHQ now which I can look at merging tomorrow?

mattigruener commented:
No problem! And yes - will do. Thanks :)

mattigruener commented:
Closing as this PR has been merged into Gaffer.
